File: est-modules/est-kotlin-support/est-go-sdk/tests/client_test.go
package est

import (
	"testing"
	"time"
)

func TestClientInitialization(t *testing.T) {
	client := NewClient("http://localhost:8080")
	if client == nil {
		t.Fatal("Expected client to be created")
	}
}

func TestClientWithOptions(t *testing.T) {
	options := ClientOptions{
		APIKey:     "test-api-key",
		Timeout:    60 * time.Second,
		MaxRetries: 5,
	}
	client := NewClientWithOptions("http://localhost:8080", options)
	if client == nil {
		t.Fatal("Expected client to be created with options")
	}
}

func TestClientDefaultOptions(t *testing.T) {
	options := ClientOptions{}
	client := NewClientWithOptions("http://localhost:8080", options)
	if client == nil {
		t.Fatal("Expected client to be created with default options")
	}
}

func TestPluginMetadataCreation(t *testing.T) {
	metadata := PluginMetadata{
		PluginID:    "test-plugin",
		Name:        "Test Plugin",
		Description: "A test plugin",
		Version:     "1.0.0",
		Author:      "Test Author",
		Category:    "test",
		Tags:        []string{"test", "example"},
		License:     "MIT",
	}

	if metadata.PluginID != "test-plugin" {
		t.Errorf("Expected plugin_id to be 'test-plugin', got %s", metadata.PluginID)
	}
	if metadata.Name != "Test Plugin" {
		t.Errorf("Expected name to be 'Test Plugin', got %s", metadata.Name)
	}
}

func TestPluginMetadataWithOptionalFields(t *testing.T) {
	homepage := "https://example.com"
	repository := "https://github.com/example/plugin"
	metadata := PluginMetadata{
		PluginID:    "test-plugin",
		Name:        "Test Plugin",
		Description: "A test plugin",
		Version:     "1.0.0",
		Author:      "Test Author",
		Category:    "test",
		Homepage:    &homepage,
		Repository:  &repository,
	}

	if metadata.Homepage == nil || *metadata.Homepage != homepage {
		t.Errorf("Expected homepage to be '%s', got %v", homepage, metadata.Homepage)
	}
	if metadata.Repository == nil || *metadata.Repository != repository {
		t.Errorf("Expected repository to be '%s', got %v", repository, metadata.Repository)
	}
}

func TestPluginVersionCreation(t *testing.T) {
	version := PluginVersion{
		Version:      "1.0.0",
		ReleaseNotes: "Initial release",
		ReleasedAt:   time.Now(),
		DownloadURL:  "http://example.com/plugin.zip",
		FileSize:     1024,
	}

	if version.Version != "1.0.0" {
		t.Errorf("Expected version to be '1.0.0', got %s", version.Version)
	}
	if version.FileSize != 1024 {
		t.Errorf("Expected file_size to be 1024, got %d", version.FileSize)
	}
}

func TestPluginReviewCreation(t *testing.T) {
	title := "Great plugin!"
	updatedAt := time.Now()
	review := PluginReview{
		ReviewID:    "review-1",
		PluginID:    "test-plugin",
		UserID:      "user-1",
		UserName:    "Test User",
		Rating:      5,
		Title:       &title,
		Content:     "Excellent plugin!",
		CreatedAt:   time.Now(),
		UpdatedAt:   &updatedAt,
		HelpfulCount: 10,
	}

	if review.Rating != 5 {
		t.Errorf("Expected rating to be 5, got %d", review.Rating)
	}
	if review.HelpfulCount != 10 {
		t.Errorf("Expected helpful_count to be 10, got %d", review.HelpfulCount)
	}
}

func TestPluginSearchCriteria(t *testing.T) {
	query := "web"
	category := "web"
	author := "john"
	minRating := 4
	criteria := PluginSearchCriteria{
		Query:     &query,
		Category:  &category,
		Tags:      []string{"framework", "ui"},
		Author:    &author,
		MinRating: &minRating,
		Page:      1,
		PageSize:  20,
		SortBy:    "rating",
		SortOrder: "asc",
	}

	if criteria.Page != 1 {
		t.Errorf("Expected page to be 1, got %d", criteria.Page)
	}
	if criteria.PageSize != 20 {
		t.Errorf("Expected page_size to be 20, got %d", criteria.PageSize)
	}
}

func TestPluginSearchResult(t *testing.T) {
	result := PluginSearchResult{
		Plugins: []PluginMetadata{
			{
				PluginID: "plugin-1",
				Name:     "Plugin 1",
			},
			{
				PluginID: "plugin-2",
				Name:     "Plugin 2",
			},
		},
		Total:    2,
		Page:     1,
		PageSize: 20,
	}

	if result.Total != 2 {
		t.Errorf("Expected total to be 2, got %d", result.Total)
	}
	if len(result.Plugins) != 2 {
		t.Errorf("Expected 2 plugins, got %d", len(result.Plugins))
	}
}

func TestPluginPublishRequest(t *testing.T) {
	homepage := "https://example.com"
	request := PluginPublishRequest{
		Name:        "New Plugin",
		Description: "A new plugin",
		Version:     "1.0.0",
		Author:      "Test Author",
		Category:    "test",
		Tags:        []string{"new", "plugin"},
		License:     "MIT",
		Homepage:    &homepage,
		FileData:    []byte("plugin-content"),
	}

	if request.Name != "New Plugin" {
		t.Errorf("Expected name to be 'New Plugin', got %s", request.Name)
	}
	if len(request.FileData) != 14 {
		t.Errorf("Expected file_data length to be 14, got %d", len(request.FileData))
	}
}

func TestPluginPublishResult(t *testing.T) {
	pluginID := "new-plugin"
	result := PluginPublishResult{
		Success:  true,
		PluginID: &pluginID,
		Message:  "Plugin published successfully",
		Warnings: []string{"warning-1", "warning-2"},
	}

	if !result.Success {
		t.Errorf("Expected success to be true, got false")
	}
	if result.PluginID == nil || *result.PluginID != pluginID {
		t.Errorf("Expected plugin_id to be '%s', got %v", pluginID, result.PluginID)
	}
	if len(result.Warnings) != 2 {
		t.Errorf("Expected 2 warnings, got %d", len(result.Warnings))
	}
}

func TestEstError(t *testing.T) {
	err := &EstError{
		Code:    "NOT_FOUND",
		Message: "Resource not found",
	}

	if err.Error() != "Resource not found" {
		t.Errorf("Expected error message to be 'Resource not found', got %s", err.Error())
	}
}

func TestEstErrorWithDetails(t *testing.T) {
	details := map[string]interface{}{
		"resource_id": "123",
		"path":        "/api/test",
	}
	err := &EstError{
		Code:    "VALIDATION_ERROR",
		Message: "Validation failed",
		Details: details,
	}

	if err.Code != "VALIDATION_ERROR" {
		t.Errorf("Expected code to be 'VALIDATION_ERROR', got %s", err.Code)
	}
	if err.Details == nil {
		t.Errorf("Expected details to be present")
	}
	if err.Details["resource_id"] != "123" {
		t.Errorf("Expected resource_id to be '123', got %v", err.Details["resource_id"])
	}
}

File: est-modules/est-kotlin-support/est-typescript-sdk/tests/client.test.ts
import { EstClient, PluginMarketplaceClient } from '../src';
import { PluginInfo, PluginReview, SearchOptions, PublishResult, EstClientConfig } from '../src/types';

describe('EstClient', () => {
  test('should initialize with base URL', () => {
    const config: EstClientConfig = { baseUrl: 'http://localhost:8080' };
    const client = new EstClient(config);
    expect(client).toBeDefined();
  });

  test('should initialize with API key', () => {
    const config: EstClientConfig = { 
      baseUrl: 'http://localhost:8080', 
      apiKey: 'test-api-key' 
    };
    const client = new EstClient(config);
    expect(client).toBeDefined();
  });

  test('should initialize with custom timeout', () => {
    const config: EstClientConfig = { 
      baseUrl: 'http://localhost:8080', 
      timeout: 60000 
    };
    const client = new EstClient(config);
    expect(client).toBeDefined();
  });

  test('should initialize with custom headers', () => {
    const config: EstClientConfig = { 
      baseUrl: 'http://localhost:8080', 
      headers: { 
        'X-Custom-Header': 'value',
        'X-Another-Header': 'another-value'
      }
    };
    const client = new EstClient(config);
    expect(client).toBeDefined();
  });

  test('should initialize with all options', () => {
    const config: EstClientConfig = { 
      baseUrl: 'http://localhost:8080', 
      apiKey: 'test-api-key',
      timeout: 60000,
      headers: { 'X-Custom': 'value' }
    };
    const client = new EstClient(config);
    expect(client).toBeDefined();
  });
});

describe('PluginMarketplaceClient', () => {
  test('should initialize with EstClient', () => {
    const client = new EstClient({ baseUrl: 'http://localhost:8080' });
    const marketplaceClient = new PluginMarketplaceClient(client);
    expect(marketplaceClient).toBeDefined();
  });

  test('should initialize with base URL string', () => {
    const marketplaceClient = new PluginMarketplaceClient('http://localhost:8080');
    expect(marketplaceClient).toBeDefined();
  });

  test('should initialize with base URL and API key', () => {
    const marketplaceClient = new PluginMarketplaceClient('http://localhost:8080', 'test-api-key');
    expect(marketplaceClient).toBeDefined();
  });
});

describe('Type Definitions', () => {
  test('should create valid PluginInfo', () => {
    const plugin: PluginInfo = {
      name: 'Test Plugin',
      version: '1.0.0',
      description: 'A test plugin',
      author: 'Test Author',
      mainClass: 'com.example.TestPlugin',
      dependencies: ['dep1', 'dep2'],
      softDependencies: ['soft-dep'],
      category: 'test',
      tags: ['test', 'example'],
      license: 'MIT',
      rating: 4.5,
      downloadCount: 1000,
      screenshots: ['screenshot1.jpg'],
      publishTime: Date.now(),
      lastUpdateTime: Date.now(),
      certified: true,
      compatibleVersions: ['2.3.0', '2.4.0'],
      minFrameworkVersion: '2.3.0'
    };
    expect(plugin.name).toBe('Test Plugin');
    expect(plugin.version).toBe('1.0.0');
    expect(plugin.certified).toBe(true);
  });

  test('should create PluginInfo with optional fields', () => {
    const plugin: PluginInfo = {
      name: 'Test Plugin',
      version: '1.0.0',
      description: 'A test plugin',
      author: 'Test Author',
      mainClass: 'com.example.TestPlugin',
      dependencies: [],
      softDependencies: [],
      category: 'test',
      tags: [],
      license: 'MIT',
      rating: 0,
      downloadCount: 0,
      screenshots: [],
      publishTime: Date.now(),
      lastUpdateTime: Date.now(),
      certified: false,
      compatibleVersions: [],
      minFrameworkVersion: '2.3.0',
      icon: 'icon.png',
      homepage: 'https://example.com',
      repository: 'https://github.com/example/plugin',
      changelog: 'Initial release'
    };
    expect(plugin.icon).toBe('icon.png');
    expect(plugin.homepage).toBe('https://example.com');
  });

  test('should create valid PluginReview', () => {
    const review: PluginReview = {
      id: 'review-1',
      pluginId: 'plugin-1',
      userId: 'user-1',
      userName: 'Test User',
      rating: 5,
      title: 'Great plugin!',
      content: 'Excellent plugin, works perfectly.',
      createdAt: Date.now(),
      updatedAt: Date.now(),
      helpfulCount: 10
    };
    expect(review.rating).toBe(5);
    expect(review.helpfulCount).toBe(10);
  });

  test('should create SearchOptions with all fields', () => {
    const options: SearchOptions = {
      query: 'web framework',
      category: 'web',
      tags: ['ui', 'framework'],
      page: 2,
      pageSize: 10,
      sortBy: 'rating',
      sortOrder: 'desc'
    };
    expect(options.query).toBe('web framework');
    expect(options.page).toBe(2);
    expect(options.sortBy).toBe('rating');
  });

  test('should create SearchOptions with minimal fields', () => {
    const options: SearchOptions = {};
    expect(options.query).toBeUndefined();
    expect(options.page).toBeUndefined();
  });

  test('should create PublishResult with success', () => {
    const result: PublishResult = {
      success: true,
      pluginId: 'new-plugin',
      version: '1.0.0',
      message: 'Plugin published successfully'
    };
    expect(result.success).toBe(true);
    expect(result.pluginId).toBe('new-plugin');
  });

  test('should create PublishResult with errors', () => {
    const result: PublishResult = {
      success: false,
      pluginId: 'failed-plugin',
      version: '1.0.0',
      message: 'Validation failed',
      errors: ['Invalid version', 'Missing description']
    };
    expect(result.success).toBe(false);
    expect(result.errors?.length).toBe(2);
  });

  test('should create EstClientConfig with minimal config', () => {
    const config: EstClientConfig = {
      baseUrl: 'http://localhost:8080'
    };
    expect(config.baseUrl).toBe('http://localhost:8080');
  });

  test('should create EstClientConfig with full config', () => {
    const config: EstClientConfig = {
      baseUrl: 'http://localhost:8080',
      apiKey: 'test-api-key',
      timeout: 60000,
      headers: {
        'X-Custom': 'value',
        'Accept': 'application/json'
      }
    };
    expect(config.apiKey).toBe('test-api-key');
    expect(config.timeout).toBe(60000);
    expect(config.headers?.['X-Custom']).toBe('value');
  });
});
